kumascript/src/api/util.js

```diff
@@ -4,22 +4,67 @@
+const cssesc = require("cssesc");
+const sanitizeFilename = require("sanitize-filename");
 
-const config = require("../config.js");
 const cheerio = require("../monkeypatched-cheerio.js");
 
-const URL_UNSAFE = /["#$%&+,/:;=?@\[\]^`{}|~'()]/g;
-
 const H1_TO_H6_TAGS = new Set(["h1", "h2", "h3", "h4", "h5", "h6"]);
 const HEADING_TAGS = new Set([...H1_TO_H6_TAGS, "hgroup"]);
 const INJECT_SECTION_ID_TAGS = new Set([...HEADING_TAGS, "section"]);
+const LIVE_SAMPLE_PARTS = ["html", "css", "js"];
+const SECTION_ID_DISALLOWED = /["#$%&+,/:;=?@[\]\^`{|}~')(\\]/g;
+
+class KumascriptError extends Error {
+  constructor(message) {
+    super(message);
+    this.name = this.constructor.name;
+  }
+}
 
 function slugify(text) {
-  // Turn the text content of a header into a slug for use in an ID.
-  // Remove unsafe characters, and collapse whitespace gaps into
-  // underscores.
-  return text.replace(URL_UNSAFE, "").trim().replace(/\s+/g, "_");
+  // Turn the text content of a header, or the value of the "name" attribute,
+  // into a slug for use as an ID as well as a filename. Trim it, collapse
+  // whitespace gaps into underscores, remove the same special characters that
+  // Kuma removes (for consistency, since for example many live-samples depend
+  // on this), lowercase it, and finally remove any remaining characters that
+  // would not work within a filename on a Windows, Mac OS X, or Unix filesystem.
+  // NOTE: These are the constraints that have to be satisfied:
+  //    1) the result can be used as a filename on Windows, Mac OS X, and Unix
+  //       (this is why the "sanitize-filename" npm package is used, and since
+  //       all of Yari's filenames are lowercase, this is why "toLowerCase" is
+  //       used)
+  //    2) the result will be used as an "id" attribute, so in HTML5 it must
+  //       contain at least one character and must not contain any whitespace
+  //       characters (the "sanitize-filename" npm package will itself remove
+  //       spaces, but since they're useful in breaking up phrases, before we
+  //       run "sanitize-filename" we convert whitespace gaps into underscores)
+  //    3) many macros use sample ID's that assume that "id" attributes have
+  //       had the SECTION_ID_DISALLOWED characters removed, so for now we have
+  //       to maintain that legacy
+  //    4) there's no need to add constraints that assume the result will be
+  //       used as a CSS ID selector, since it will be properly escaped for that
+  //       use case (see the "cssesc" code within the "getSection" method of the
+  //       HTMLTool below)
+  return sanitizeFilename(
+    text
+      .trim()
+      .replace(/\s+/g, "_")
+      .replace(SECTION_ID_DISALLOWED, "")
+      .toLowerCase()
+  );
+}
+
+function spacesToUnderscores(text) {
+  return text.replace(/ |%20/g, "_");
+}
+
+function safeDecodeURIComponent(text) {
+  try {
+    return decodeURIComponent(text);
+  } catch (e) {
+    return text;
```

Comment:
What does this mean?! Why is OK for `decodeURIComponent` to fail?

Reply:
Some of the macro calls to `EmbedLiveSample` URI-encode the sample ID in anticipation of it being used as part of the live sample URL (I think this happens only on localized pages currently). For example, consider https://wiki.developer.mozilla.org/zh-CN/docs/Web/API/Canvas_API/Tutorial/Basic_usage?raw which makes this call:
```
{{EmbedLiveSample("%E4%B8%80%E4%B8%AA%E6%A8%A1%E6%9D%BF%E9%AA%A8%E6%9E%B6", 160, 160)}}
```
but could have and should have just done this:
```
{{EmbedLiveSample("一个模板骨架", 160, 160)}}
```
since the sample ID of "一个模板骨架" will be URI-encoded within the macro itself.

So my `safeDecodeURIComponent()` function simply attempts to decode its input not knowing if its URI encoded or not, and if it fails to decode, I assume it's not encoded (as most sample ID's aren't) and move on.

I added this only to handle the sloppiness of the localized content. The English documents don't ever do this.

Reply:
I added a comment to the function to explain this.